### PR TITLE
Compute model size callback

### PR DESCRIPTION
**Pull request by samadejacobs**
Callback to calculate model size at specified batch interval.

### CODE REVIEW

**Review by timmoon10 — CHANGES_REQUESTED** [review]
I don't think this approach will work. This manually sets the activations of a constant layer, but the value will be overwritten during forward prop. Possible solutions:
1. All the parameter sizes are known at model construction time, so the best approach would be to construct constant layers with the correct values in the Python front-end. If you have sub-models that you want to mix and match, then you could have multiple constant layers:
```python
gen = MyGeneratorModule()
disc = MyDiscriminatorModule()
real_pred = disc(lbann.Input())
fake_pred = disc(gen())
gen_size = lbann.Constant(value=str(gen.num_params), num_neurons='1')
disc_size = lbann.Constant(value=str(disc.num_params), num_neurons='1')
total_size = lbann.Add(gen_size, disc_size)
```
2. If we want to dynamically compute the model size, it's probably best to store the value inside a weights object rather than a constant layer. This will have a nicer implementation since weights exposes an API to access the value. The result would look like:
```python
size_weights = lbann.Weights(optimizer=lbann.NoOptimizer(), name='num_params')
size_layer = lbann.WeightsLayer(weights=size_weights, dims='1')
size_callback = lbann.CallbackComputeModelSize(weights='num_params', batch_interval=123)
```
3. The callback can modify the `m_value` member of the constant layer, either directly or by adding an access function to the constant layer. This will be ugly.

File: include/lbann/callbacks/compute_model_size.hpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#include "lbann/layers/transform/constant.hpp"
+
+
+namespace lbann {
+namespace callback {
+
+/** @brief Compute model (parameter size).
+ *
+ *  Goes through learning layers, compute size and 
+ *  write computed valude to output layer
```
[the remark this answers is not in the document]
Comment (samadejacobs) [reply include/lbann/callbacks/compute_model_size.hpp]: Done

File: include/lbann/callbacks/compute_model_size.hpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+
+namespace lbann {
+namespace callback {
+
+/** @brief Compute model (parameter size).
+ *
+ *  Goes through learning layers, compute size and 
+ *  write computed value to named output weight layer
+ */
```
Comment (timmoon10) [inline include/lbann/callbacks/compute_model_size.hpp:41]:
```suggestion
/** @brief Compute number of parameters in model.
 *
 *  The model size is output into a user-provided weights object, 
 *  which is expected to have one entry.
 */
```

**Review by timmoon10 — APPROVED** [review]
LGTM

**Pull request merged**